### ISSUE TITLE
bug反馈：推理阶段错误调用encoder

### ISSUE THREAD

**Issue by SuMeng123** [THU-BPM/PairSCL#6]
https://github.com/THU-BPM/PairSCL/blob/5f2dcf042bb3c2d9b5fb90573af518d25b1ea73d/main_test.py#L79
model.encoder(**inputs)是直接计算bert的表征了，感觉是一个bug，需要改为以下这样。我在SNLI任务上实际跑了下代码，修改前ACC=33%，修改后ACC=86.81%。至于为什么ACC没有像论文的结果那样，我还在找原因，但是起码证明这行代码是有问题。这个错误同样存在于main_validate.py的validate方法
` features = model(**inputs)`
` logits = classifier(features.detach())`

**Comment by thulishuang** [comment]
这里确实写错了，features应该是通过model计算得到

**Comment by cingtiye** [comment]
> https://github.com/THU-BPM/PairSCL/blob/5f2dcf042bb3c2d9b5fb90573af518d25b1ea73d/main_test.py#L79
> 
> 
> model.encoder(**inputs)是直接计算bert的表征了，感觉是一个bug，需要改为以下这样。我在SNLI任务上实际跑了下代码，修改前ACC=33%，修改后ACC=86.81%。至于为什么ACC没有像论文的结果那样，我还在找原因，但是起码证明这行代码是有问题。这个错误同样存在于main_validate.py的validate方法
> ` features = model(**inputs)`
> ` logits = classifier(features.detach())`

你好，请问你后面复现了论文的结果了吗？

**Comment by SuMeng123** [comment]
> 

即使修改了本issue中提到的问题，依然没有复现成功，在SNLI任务上ACC=86.81%。我尝试自己寻找原因，发现了很多不解的地方：
1、在main_supcon.py计算交叉熵loss的时候，criterion_ce(features, batch[3]) 中的features维度是[batchsize, 128]，感觉是不合理的吧？这样感觉无法计算交叉熵。应该先将features维度变为[batchsize, 3]之后，再计算criterion_ce才有意义
2、在main_supcon.py中，loss = criterion_sup(features, batch[3]) + criterion_ce(features, batch[3]) ，我尝试把criterion_ce移除，结果发现训练loss几乎没有下降，这点没想明白原因。
3、复现过程中，我的batch_size设置的128，学习率0.00005，由于出现训练loss为nan，所以我增加了梯度截断，其余代码没有变动。为什么loss为nan我不清楚，难道是bs太低？请问你们实验中的bs是多少，与论文（512）一致吗

**Comment by cingtiye** [comment]
> > 
> 
> 即使修改了本issue中提到的问题，依然没有复现成功，在SNLI任务上ACC=86.81%。我尝试自己寻找原因，发现了很多不解的地方： 1、在main_supcon.py计算交叉熵loss的时候，criterion_ce(features, batch[3]) 中的features维度是[batchsize, 128]，感觉是不合理的吧？这样感觉无法计算交叉熵。应该先将features维度变为[batchsize, 3]之后，再计算criterion_ce才有意义 2、在main_supcon.py中，loss = criterion_sup(features, batch[3]) + criterion_ce(features, batch[3]) ，我尝试把criterion_ce移除，结果发现训练loss几乎没有下降，这点没想明白原因。 3、复现过程中，我的batch_size设置的128，学习率0.00005，由于出现训练loss为nan，所以我增加了梯度截断，其余代码没有变动。为什么loss为nan我不清楚，难道是bs太低？请问你们实验中的bs是多少，与论文（512）一致吗

还在跑~我用四张卡，batch_size=512，学习率5e-5，跑完再告诉你~

**Comment by cingtiye** [comment]
> > 
> 
> 即使修改了本issue中提到的问题，依然没有复现成功，在SNLI任务上ACC=86.81%。我尝试自己寻找原因，发现了很多不解的地方： 1、在main_supcon.py计算交叉熵loss的时候，criterion_ce(features, batch[3]) 中的features维度是[batchsize, 128]，感觉是不合理的吧？这样感觉无法计算交叉熵。应该先将features维度变为[batchsize, 3]之后，再计算criterion_ce才有意义 2、在main_supcon.py中，loss = criterion_sup(features, batch[3]) + criterion_ce(features, batch[3]) ，我尝试把criterion_ce移除，结果发现训练loss几乎没有下降，这点没想明白原因。 3、复现过程中，我的batch_size设置的128，学习率0.00005，由于出现训练loss为nan，所以我增加了梯度截断，其余代码没有变动。为什么loss为nan我不清楚，难道是bs太低？请问你们实验中的bs是多少，与论文（512）一致吗

跑完和你一样~
我觉得这个代码上传之前作者应该是没有测试过得~
另外这个代码的输入我觉得都不对
https://github.com/THU-BPM/PairSCL/blob/main/bert_model.py#:~:text=input_ids2%20%3D%20input_ids,input_ids%20%2D%20input_ids2
你品一下这个输入~你细品~
